File: app/src/main/java/xyz/zpayh/myadapter/ExpandableActivity.java
package xyz.zpayh.myadapter;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SimpleItemAnimator;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import xyz.zpayh.adapter.DefaultExpandable;
import xyz.zpayh.adapter.DiffUtilCallback;
import xyz.zpayh.adapter.IExpandable;
import xyz.zpayh.adapter.IMultiItem;
import xyz.zpayh.adapter.OnItemClickListener;
import xyz.zpayh.adapter.OnItemLongClickListener;
import xyz.zpayh.myadapter.adapter.MyExpandableAdapter;
import xyz.zpayh.myadapter.data.Card;
import xyz.zpayh.myadapter.data.ImageCard;
import xyz.zpayh.myadapter.data.ImageLabel;

public class ExpandableActivity extends AppCompatActivity {

    private String mLabels[] = {"Animal","Beauty","Scenery","Tranquil"};

    // Animal Card
    private int mAnimalIds[] = {R.drawable.fuchs,R.drawable.horses};
    private String mAnimalTitles[] = {"Fuchs","Horses"};

    // Beauty Card
    private int mBeautyIds[] = {R.drawable.adult, R.drawable.girl,
            R.drawable.girl1,R.drawable.girl2,R.drawable.smile,R.drawable.fashion};
    private String mBeautyTitles[] = {"Adult","Girl", "Girl MM",
            "MM Girl","Smile","fashion"};

    // Scenery Card
    private int mSceneryIds[] = {R.drawable.architecture,R.drawable.denmark,R.drawable.sunset};
    private String mSceneryTitles[] = {"Architecture","Denmark","Sunset"};

    // Tranquil Card
    private int mTranquilIds[] = {R.drawable.easter_eggs,R.drawable.notes};
    private String mTranquilTitles[] = {"Easter Eggs","Notes"};

    private RecyclerView mRecyclerView;
    private MyExpandableAdapter mAdapter;
    private GridLayoutManager mGridLayoutManager;
    private StaggeredGridLayoutManager mStaggeredGridLayoutManager;

    private boolean mIsGrid = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_expandable);
        initView();
    }

    private void initView() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        mRecyclerView = (RecyclerView) findViewById(R.id.list);
        mGridLayoutManager = new GridLayoutManager(this,2);
        mStaggeredGridLayoutManager = new StaggeredGridLayoutManager(2,StaggeredGridLayoutManager.VERTICAL);
        mRecyclerView.setLayoutManager(mStaggeredGridLayoutManager);
        mAdapter = new MyExpandableAdapter();

        //mAdapter.setAlwaysShowFoot(true);
        //mAdapter.setAlwaysShowHead(true);

        mAdapter.addHeadLayout(R.layout.item_head);
        mAdapter.addFootLayout(R.layout.item_foot2);

        // 设置DiffUtil效果
        mAdapter.setCallback(new DiffUtilCallback<IMultiItem>() {
            @Override
            public boolean areItemsTheSame(IMultiItem oldItem, IMultiItem newItem) {
                if (oldItem instanceof ImageLabel && newItem instanceof ImageLabel){
                    return TextUtils.equals(((ImageLabel) oldItem).getData(),
                            ((ImageLabel) newItem).getData());
                }
                if (oldItem instanceof Card && newItem instanceof Card){
                    return ((Card) oldItem).getData().mImageResId == ((Card) newItem).getData().mImageResId;
                }
                return false;
            }

            @Override
            public boolean areContentsTheSame(IMultiItem oldItem, IMultiItem newItem) {
                if (oldItem instanceof ImageLabel && newItem instanceof ImageLabel){
                    return true;
                }
                if (oldItem instanceof Card && newItem instanceof Card){
                    return TextUtils.equals(((Card) oldItem).getData().mImageTitle,
                            ((Card) newItem).getData().mImageTitle);
                }
                return false;
            }
        });

        mRecyclerView.setAdapter(mAdapter);
        //屏障默认的Change动画
        ((SimpleItemAnimator)mRecyclerView.getItemAnimator()).setSupportsChangeAnimations(false);

        // 点击展开/折叠事件
        mAdapter.setOnItemClickListener(new OnItemClickListener() {
            @Override
            public void onItemClick(@NonNull View view, int adapterPosition) {
                if (view.getId() == R.id.label_root){
                    IMultiItem item = mAdapter.getData(adapterPosition);
                    if (item instanceof IExpandable){
                        IExpandable expandable = (IExpandable) item;
                        if (expandable.isExpandable()){
                            mAdapter.collapseAll(adapterPosition);
                        }else{
                            mAdapter.expandAll(adapterPosition);
                        }
                    }
                }
            }
        });

        mAdapter.setOnItemLongClickListener(new OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(@NonNull View view, int adapterPosition) {
                mAdapter.removeData(adapterPosition);
                return true;
            }
        });

        final SwipeRefreshLayout refresh = (SwipeRefreshLayout) findViewById(R.id.refresh);

        refresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                refreshData();
                refresh.setRefreshing(false);
            }
        });

        findViewById(R.id.fab).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                refreshData();
            }
        });

        initData();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_grid:
                if (mIsGrid){
                    return true;
                }
                mIsGrid = true;
                mRecyclerView.setLayoutManager(mGridLayoutManager);
                mAdapter.onAttachedToRecyclerView(mRecyclerView);
                initData();
                return true;
            case R.id.action_staggered:
                if (!mIsGrid){
                    return true;
                }
                mIsGrid = false;
                mRecyclerView.setLayoutManager(mStaggeredGridLayoutManager);
                initData();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.multiitem_menu, menu);
        return true;
    }

    private void refreshData(){

        List<IMultiItem> oldData = mAdapter.getData();
        List<IMultiItem> data = new ArrayList<>();
        data.addAll(oldData);

        Collections.reverse(data);

        DefaultExpandable expandable = (DefaultExpandable) data.get(1);
        Collections.reverse(expandable.getSubItems());
        expandable = (DefaultExpandable) data.get(2);
        Collections.reverse(expandable.getSubItems());

        mAdapter.setData(data);
    }

    private void initData(){
        int width = getResources().getDisplayMetrics().widthPixels / mStaggeredGridLayoutManager.getSpanCount();

        List<IMultiItem> data = new ArrayList<>();

        for (int i = 0; i < mLabels.length; i++) {
            ImageLabel label = new ImageLabel(mLabels[i]);
            data.add(label);
        }

        // Add animals
        List<IMultiItem> animalData = new ArrayList<>(2);
        for (int i = 0; i < mAnimalIds.length; i++) {
            ImageCard cardData = new ImageCard(mAnimalIds[i],mAnimalTitles[i]);
            Card card = new Card(width,cardData);
            animalData.add(card);
        }
        ((DefaultExpandable)data.get(0)).addSubData(animalData);

        // Add beauties
        List<IMultiItem> beautyData = new ArrayList<>(mBeautyIds.length);
        for (int i = 0; i < mBeautyIds.length; i++) {
            ImageCard cardData = new ImageCard(mBeautyIds[i],mBeautyTitles[i]);
            Card card = new Card(width,cardData);
            beautyData.add(card);
        }
        ((DefaultExpandable)data.get(1)).addSubData(beautyData);

        // Add scenery
        List<IMultiItem> sceneryData = new ArrayList<>(mSceneryIds.length);
        for (int i = 0; i < mSceneryIds.length; i++) {
            ImageCard cardData = new ImageCard(mSceneryIds[i],mSceneryTitles[i]);
            Card card = new Card(width,cardData);
            sceneryData.add(card);
        }
        ((DefaultExpandable)data.get(2)).addSubData(sceneryData);

        // Add tranquil
        List<IMultiItem> tranquilData = new ArrayList<>(mTranquilIds.length);
        for (int i = 0; i < mTranquilIds.length; i++) {
            ImageCard cardData = new ImageCard(mTranquilIds[i],mTranquilTitles[i]);
            Card card = new Card(width,cardData);
            tranquilData.add(card);
        }
        ((DefaultExpandable)data.get(3)).addSubData(tranquilData);

        mAdapter.setData(data);
    }
}

File: sample-kotlin/src/main/java/cn/sherlockzp/sample/ExpandableActivity.kt
package cn.sherlockzp.sample

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.util.DiffUtil
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.SimpleItemAnimator
import android.support.v7.widget.StaggeredGridLayoutManager
import android.view.Menu
import android.view.MenuItem
import cn.sherlockzp.adapter.*
import cn.sherlockzp.sample.model.Card
import cn.sherlockzp.sample.model.ImageLabel

import kotlinx.android.synthetic.main.activity_expandable.*

class ExpandableActivity : AppCompatActivity() {

    private val labels = arrayOf("Animal","Beauty","Scenery","Tranquil")

    // Animal Card
    private val animalIds = arrayOf(R.drawable.fuchs,R.drawable.horses)
    private val animalTitles = arrayOf("Fuchs","Horses")

    //Beauty Card
    private val beautyIds = arrayOf(R.drawable.adult, R.drawable.girl,
            R.drawable.girl1,R.drawable.girl2,R.drawable.smile,R.drawable.fashion)
    private val beautyTitles = arrayOf("Adult","Girl", "Girl MM",
            "MM Girl","Smile","fashion")

    // Scenery Card
    private val sceneryIds = arrayOf(R.drawable.architecture,R.drawable.denmark,R.drawable.sunset)
    private val sceneryTitles = arrayOf("Architecture","Denmark","Sunset")

    // Tranquil Card
    private val tranquilIds = arrayOf(R.drawable.easter_eggs,R.drawable.notes)
    private val tranquilTitle = arrayOf("Easter Eggs","Notes")

    private val adapter = object : BaseExpandableAdapter(){
        override fun convertHead(holder: BaseViewHolder, headLayout: LayoutConfig) {
            holder.setText(R.id.tv_head, "这是图片展")
        }
    }

    private val gridLayoutManager by lazy {
        GridLayoutManager(this,2)
    }
    private val staggeredGridLayoutManager by lazy {
        StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
    }
    private var isGrid = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_expandable)

        initView()
        initData()
    }

    private fun initView() {
        rv_list_expand.layoutManager = staggeredGridLayoutManager
        rv_list_expand.adapter = adapter
        val animator = rv_list_expand.itemAnimator
        if (animator is SimpleItemAnimator) {
            animator.supportsChangeAnimations = false
        }

        with(adapter) {
            addHeadLayout(R.layout.item_head,spanSize = 2)
            addFootLayout(R.layout.item_foot2, spanSize = 2)
            setOnItemClickListener { view, adapterPosition ->
                if (view.id == R.id.label_root) {
                    val item = getData(adapterPosition)
                    if (item is IExpandable) {
                        if (item.expandable) {
                            // 修复issue#2
                            // 这个支持包的bug: https://issuetracker.google.com/issues/37034096
                            staggeredGridLayoutManager.invalidateSpanAssignments()
                            collapseAll(adapterPosition)
                        } else {
                            expandAll(adapterPosition)
                        }
                    }
                }
            }
            setOnItemLongClickListener { _, adapterPosition ->
                if (adapterPosition < headSize) {
                    removeHead(adapterPosition)
                    return@setOnItemLongClickListener true
                }
                if (adapterPosition >= headSize + getShowDataSize()) {
                    removeFoot(adapterPosition)
                    return@setOnItemLongClickListener true
                }
                removeDataAt(adapterPosition)
                return@setOnItemLongClickListener true
            }
            setItemDiffCallback(object : DiffUtil.ItemCallback<IMultiItem>(){
                override fun areItemsTheSame(oldItem: IMultiItem, newItem: IMultiItem): Boolean {
                    if (oldItem is ImageLabel && newItem is ImageLabel) {
                        return oldItem.label == newItem.label
                    }
                    if (oldItem is Card && newItem is Card) {
                        return oldItem.imageResId == newItem.imageResId
                    }
                    return false
                }

                override fun areContentsTheSame(oldItem: IMultiItem, newItem: IMultiItem): Boolean {
                    if (oldItem is ImageLabel && newItem is ImageLabel) {
                        return true
                    }
                    if (oldItem is Card && newItem is Card) {
                        return oldItem.title == newItem.title
                    }
                    return false
                }
            })
        }

        sr_refresh.setOnRefreshListener {
            refreshData()
            sr_refresh.isRefreshing = false
        }
    }

    private fun refreshData(){
        val oldData = adapter.data

        val newData = ArrayList(oldData).reversed()

        if (newData.size > 1) {
            val expandable = newData[1]
            if (expandable is ImageLabel){
                expandable.getSubItems()?.reverse()
            }
        }
        if (newData.size > 2) {
            val expandable = newData[2]
            if (expandable is ImageLabel){
                expandable.getSubItems()?.reverse()
            }
        }
        adapter.setData(newData)
    }

    private fun initData() {
        val width = resources.displayMetrics.widthPixels

        val data = ArrayList<IMultiItem>()
        labels.forEach {
            data.add(ImageLabel(it))
        }

        // Add Animals
        val animalData = ArrayList<IMultiItem>(2)
        for (i in animalIds.indices) {
            animalData.add(Card(width,animalIds[i],animalTitles[i]))
        }
        (data[0] as ImageLabel).subItems.addAll(animalData)

        // Add Beauties
        val beautyData = ArrayList<IMultiItem>(beautyIds.size)
        for (i in beautyIds.indices) {
            beautyData.add(Card(width, beautyIds[i], beautyTitles[i]))
        }
        (data[1] as ImageLabel).subItems.addAll(beautyData)

        // Add scenery
        val sceneryData = ArrayList<IMultiItem>(sceneryIds.size)
        for (i in sceneryIds.indices) {
            sceneryData.add(Card(width, sceneryIds[i], sceneryTitles[i]))
        }
        (data[2] as ImageLabel).subItems.addAll(sceneryData)

        // Add tranquil
        val tranquilData = ArrayList<IMultiItem>(tranquilIds.size)
        for (i in tranquilIds.indices) {
            tranquilData.add(Card(width, tranquilIds[i], tranquilTitle[i]))
        }
        (data[3] as ImageLabel).subItems.addAll(animalData)

        adapter.setData(data)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when(item?.itemId) {
            R.id.action_grid -> {
                if (isGrid) {
                    return true
                }
                isGrid = true
                rv_list_expand.layoutManager = gridLayoutManager
                adapter.onAttachedToRecyclerView(rv_list_expand)
                initData()
                return true
            }
            R.id.action_staggered -> {
                if (!isGrid) {
                    return true
                }
                isGrid = false
                rv_list_expand.layoutManager = staggeredGridLayoutManager
                adapter.onAttachedToRecyclerView(rv_list_expand)
                initData()
                return true
            }
            R.id.action_clear -> adapter.setData(null)
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.multiitem_menu, menu)
        return true
    }
}
